### PR TITLE
Clean blockchain usage of cryptographic key-pairs

### LINKED ISSUE #475: Simplify blockchain example to only use Edward's keys
Simplify blockchain example to only use Edward's keys

_Originally posted by @prince-chrismc in https://github.com/pyrsia/pyrsia/pull/461#discussion_r822127103_

### Context

Summary is here https://github.com/pyrsia/pyrsia/issues/475#issuecomment-1063098820

https://github.com/pyrsia/pyrsia/blob/95badf1e55c4373e4dee184ac2f64debcca68d41/src/blockchain/examples/simple_node.rs#L66-L70

### Goal

We should  only use ` identity::Keypair::Ed25519`

### PR DESCRIPTION
**Pull request by prince-chrismc**
## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

## Description

Fixes #475 

This PR cleans out the usage of the key pairs. The blockchain is only supporting edwards keys are of now.

## Extras

- Add the blockchain crate to the root toml so the deps are in sync
- Add the example to the root and make sure it's build in CI
- clean up the rust docs values (it duplicated the field name)
- fixup clippy warnings

### CODE REVIEW

**Comment by prince-chrismc** [comment]
@sbtaylor15 I added you so can follow the changes to the rust workflow, this impacts #451 if it goes in first

File: Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ systemd-units = { unit-name = "pyrsia", unit-scripts = "installers/debian/" }
 [dependencies]
 signed = {path="src/signed"}
 signed_struct = { path = "src/signed_struct" }
+pyrsia_blockchain_network = { path = "src/blockchain" }
```
Comment (chb0github) [inline Cargo.toml:25]: yes!

**Review by chb0github — DISMISSED** [review]

**Comment by prince-chrismc** [comment]
CI broke 😕

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -79,8 +79,8 @@ jobs:
 
     - name: Build Release Binaries
       run: |
-        cargo build --workspace --verbose  --release
-    
+        cargo build --workspace --examples --release
```
Comment (tiainen) [inline .github/workflows/rust.yml:82]: If you [specify a target with cargo build](https://doc.rust-lang.org/cargo/commands/cargo-build.html#target-selection), it will only build that target. While leaving out any target selection will build all targets, including any examples. This is also the reason why the bins are no longer being built, which causes the failure during the deb package generation.

I would suggest to just keep the `--verbose` option and leave out any specific target to build all targets.

Comment (tiainen) [reply .github/workflows/rust.yml:82]: I was wrong with the assumption that all targets are built when specifying no targets. By default, it builds all binary and library targets. If the examples need to be included, specify `--lib --bins --examples` (or `--all-targets` which includes tests and benchmarks as well).

**Review by fishseabowl — APPROVED** [review]
it looks good to me

File: .github/workflows/rust.yml (hunk 2 above)
Comment (prince-chrismc) [reply .github/workflows/rust.yml]: Thanks for looking up the docs ❤️ let's see if it passes now 🤞

**Comment by prince-chrismc** [comment]
thanks @tiainen, you found a working solution 🚀 

😻

**Review by tiainen — DISMISSED** [review]

**Review by fishseabowl — DISMISSED** [review]
it looks fine to me.

**Review by chb0github — DISMISSED** [review]

File: src/blockchain/src/header.rs [hunk 3]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
 /// struct Header define the header of a block
 #[derive(Serialize, Deserialize, Debug, Clone, Hash, PartialEq, Eq, Copy)]
 pub struct Header {
-    ///      parent_hash: 256bit Keccak Hash of the Parent Block(previous Block hash)
+    /// 256bit Keccak Hash of the Parent Block(previous Block hash)
     pub parent_hash: HashDigest,
-    ///      committer:  the committer node's PeerID
+    /// the committer node's PeerID
```
Comment (betarelease) [inline src/blockchain/src/header.rs:30]: can these comments be made part of the naming of the variable?
For example in this case

renaming commiter - to committerPeerId would convey what it is and not require the comment.

File: src/blockchain/src/header.rs [hunk 4]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
...
     pub committer: Address,
-    ///      transactions_root: 256bit Keccak Hash of the root node of Transactions Merkle tree
+    /// 256bit Keccak Hash of the root node of Transactions Merkle tree
     pub transactions_root: HashDigest,
-    ///      timestamp: Unix tim, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
+    /// Unix time, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
```
Comment (betarelease) [inline src/blockchain/src/header.rs:34]: This comment seems to not provide any extra information. By convention timestamps are usually since epoch and either they are seconds or milliseconds.
This can be renamed timestampInsecs so that the comment can be removed.

File: src/blockchain/src/header.rs [hunk 5]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
...
     pub timestamp: u64,
-    ///      number: block sequence number, the current block number should be the parent(previous) block number plus 1
+    /// block sequence number, the current block number should be the parent(previous) block number plus 1
```
Comment (betarelease) [inline src/blockchain/src/header.rs:36]: number is very general and does not tell me whether it is a whole number or integer or any other type of number.

Since this is the blockSequenceNumber lets name it as such.

File: src/blockchain/src/header.rs [hunk 6]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
...
     pub number: u128,
-    nonce: u128, // Adds a salt to harden
-    ///      hash: block id, 256bit Keccak Hash of the Current Block Header, excluding itself
+    /// Adds a salt to harden
+    nonce: u128,
+    /// block id, 256bit Keccak Hash of the Current Block Header, excluding itself
```
Comment (betarelease) [inline src/blockchain/src/header.rs:40]: Are we using the 256bit Keccak Hash whenever we mean HashDigest? In that case this comment can be removed as well.

File: src/blockchain/src/header.rs [hunk 7]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
...
     pub transactions_root: HashDigest,
-    ///      timestamp: Unix tim, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
+    /// Unix time, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
     pub timestamp: u64,
-    ///      number: block sequence number, the current block number should be the parent(previous) block number plus 1
+    /// block sequence number, the current block number should be the parent(previous) block number plus 1
     pub number: u128,
-    nonce: u128, // Adds a salt to harden
-    ///      hash: block id, 256bit Keccak Hash of the Current Block Header, excluding itself
+    /// Adds a salt to harden
```
Comment (betarelease) [inline src/blockchain/src/header.rs:38]: nonce explains itself. So this comment does not add any value.

File: .github/workflows/rust.yml [hunk 8]
```diff
@@ -79,8 +79,8 @@ jobs:
 
     - name: Build Release Binaries
       run: |
-        cargo build --workspace --verbose  --release
-    
+        cargo build --workspace --verbose --all-targets --release
```
Comment (betarelease) [inline .github/workflows/rust.yml:82]: why did we need to add --all-targets. I looked at the commit message and all it told me was we were trying with all-targets. But why was this change needed? Can you change the commit message to reflect why..

Comment (prince-chrismc) [reply .github/workflows/rust.yml:82]: > Add the example to the root and make sure it's build in CI

I'll pretty the squash commit as always

File: src/blockchain/src/header.rs (hunk 7 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:38]: It's required since it can be confused for those thinking Proof of work might be mislead of why this is here

File: src/blockchain/src/header.rs (hunk 6 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:40]: HashDigest can be any hash format, for now it's limited in our usage

File: src/blockchain/src/header.rs (hunk 5 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:36]: that's a way too long name 🙈 also it will be a bunch of refactoring an impact other issues #357 for example
or #322 needs it too

File: src/blockchain/src/header.rs (hunk 3 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:30]: existing issue #486

File: src/blockchain/src/header.rs (hunk 4 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:34]: Others might not have the same assumption as you when you think of "time stamps", it does not harm any one.

This is the only comment that could be actions but I do want to dismiss my other reviews.

Can we compromise and leave it? Maybe comment in #413 about this?

File: src/blockchain/src/header.rs (hunk 7 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:38]: this is the meaning is the top few search results FYI https://www.google.com/search?q=blockchain+block+header+nonce

File: src/blockchain/src/header.rs (hunk 4 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:34]: Now there's merge conflicts.... I fix this

Comment (betarelease) [reply src/blockchain/src/header.rs:34]: Ok.

Comment (prince-chrismc) [reply src/blockchain/src/header.rs:34]:
```suggestion
    /// Unix timestamp, see https://en.wikipedia.org/wiki/Unix_time
```

File: src/blockchain/src/header.rs (hunk 6 above)
Comment (prince-chrismc) [reply src/blockchain/src/header.rs:40]: #413 also tackles this

**Review by fishseabowl — APPROVED** [review]
It looks fine to me

File: src/blockchain/src/header.rs (hunk 5 above)
Comment (betarelease) [reply src/blockchain/src/header.rs:36]: I still prefer descriptive names over brevity. In general I think it is time well spent figuring out good names. My suggestion is only a start. In any case if it affects too many elements we can keep it for now.

File: .github/workflows/rust.yml (hunk 8 above)
Comment (betarelease) [reply .github/workflows/rust.yml:82]: When you squash can you explain how all-targets and `Add the example to the root and make sure it's build in CI` is related

**Review by betarelease — COMMENTED** [review]
I resolved most conversations that  I could find and were open.

File: pre-commit.bat [hunk 9]
```diff
@@ -19,4 +19,4 @@ cargo fmt --check
 IF %ERRORLEVEL% NEQ 0 (ECHO Cargo format failed. &Exit /b 1)
 cargo test --workspace
 IF %ERRORLEVEL% NEQ 0 (ECHO Cargo test failed. &Exit /b 1)
-cargo build --workspace
+cargo build --workspace --all-targets
```
Comment (AbhijithGanesh) [inline pre-commit.bat:22]: changes noted👍

File: src/blockchain/src/header.rs [hunk 10]
```diff
@@ -25,18 +25,19 @@ pub type Address = HashDigest;
...
     pub parent_hash: HashDigest,
-    ///      committer:  the committer node's PeerID
+    /// the committer node's PeerID
     pub committer: Address,
-    ///      transactions_root: 256bit Keccak Hash of the root node of Transactions Merkle tree
+    /// 256bit Keccak Hash of the root node of Transactions Merkle tree
     pub transactions_root: HashDigest,
-    ///      timestamp: Unix tim, the number of seconds that have elapsed since the Unix epoch, excluding leap seconds
+    /// Unix timestamp, see https://en.wikipedia.org/wiki/Unix_time
     pub timestamp: u64,
```
Comment (AbhijithGanesh) [inline src/blockchain/src/header.rs:35]: Will we sicking to this format as there were mentions of `iso_8601` elsewhere, Is this the de-facto one?

**Review by AbhijithGanesh — APPROVED** [review]

**Review by betarelease — APPROVED** [review]

**Review by betarelease — DISMISSED** [review]

**Review by betarelease — APPROVED** [review]

**Review by fishseabowl — APPROVED** [review]
LGTM

**Review by tiainen — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Thank for taking the time to review!!!! 🎉
